mcfly.zsh

```diff
@@ -0,0 +1,103 @@
+#!/bin/zsh
+
+# Ensure stdin is a tty
+[[ -t 0 ]] || return
+
+# Avoid loading this file more than once
+if [[ "$__MCFLY_LOADED" == "loaded" ]]; then
+  return 0
+fi
+__MCFLY_LOADED="loaded"
+
+# Ensure HISTFILE exists.
+export HISTFILE="${HISTFILE:-$HOME/.zsh_history}"
+if [[ ! -r "${HISTFILE}" ]]; then
+  echo "McFly: ${HISTFILE} does not exist or is not readable. Please fix this or set HISTFILE to something else before using McFly."
+  return 1
+fi
+
+# MCFLY_SESSION_ID is used by McFly internally to keep track of the commands from a particular terminal session.
+export MCFLY_SESSION_ID=$(dd if=/dev/urandom bs=256 count=1 2> /dev/null | env LC_ALL=C tr -dc 'a-zA-Z0-9' | head -c 24)
+
+setopt extended_history       # record timestamp of command in HISTFILE
+setopt hist_expire_dups_first # delete duplicates first when HISTFILE size exceeds HISTSIZE
+setopt hist_ignore_dups       # ignore duplicated commands history list
+setopt hist_ignore_space      # ignore commands that start with space
+setopt inc_append_history     # add commands to HISTFILE in order of execution
+setopt share_history          # share command history data
+setopt interactive_comments   # allow comments in interactive shells (like Bash does)
+
+# Setup a function to clear the shell history.
+function erase_history { local HISTSIZE=0; }
+
+# Setup a function to be used by $PROMPT_COMMAND.
+function mcfly_prompt_command {
+  local exit_code=$? # Record exit status of previous command.
+
+  # Populate McFly's temporary, per-session history file from recent commands in the shell's primary HISTFILE.
+  if [[ ! -f "${MCFLY_HISTORY}" ]]; then
+    export MCFLY_HISTORY=$(mktemp -t mcfly.XXXXXXXX)
+    tail -n100 "${HISTFILE}" >| ${MCFLY_HISTORY}
+  fi
+
+  if [[ $(fc -Iln 2>/dev/null | tail -n1) = ' mcfly search'* ]] ; then
+    # If the most recent history item is a mcfly search, do not update $MCFLY_HISTORY.
+    [ -n "$MCFLY_DEBUG" ] && echo "Not appending history to $MCFLY_HISTORY."
+  else
+    # Write history to $MCFLY_HISTORY.
+    fc -W "${MCFLY_HISTORY}"
+  fi
+
+  # Run mcfly with the saved code. It will:
+  # * append commands to $HISTFILE, (~/.zsh_history by default)
+  #   for backwards compatibility and to load in new terminal sessions;
+  # * find the text of the last command in $MCFLY_HISTORY and save it to the database.
+  [ -n "$MCFLY_DEBUG" ] && echo "Run mcfly add --exit ${exit_code}"
+  mcfly add --exit ${exit_code} # --append-to-histfile --zsh-extended-history
+  # Clear the in-memory history and reload it from $MCFLY_HISTORY
+  # (to remove instances of '#mcfly: ' from the local session history).
+  erase_history
+  fc -R "${MCFLY_HISTORY}"
+  return ${exit_code} # Restore the original exit code by returning it.
+}
+
+# Set $PROMPT_COMMAND run mcfly_prompt_command and then any existing $PROMPT_COMMAND.
+PROMPT_COMMAND="mcfly_prompt_command;$PROMPT_COMMAND"
+precmd_functions+=(mcfly_prompt_command)
```

Comment:
I think `PROMPT_COMMAND` is bash only?
BTW, you can run a command on every prompt in zsh robustly like this:
```zsh
autoload -Uz add-zsh-hook
add-zsh-hook preexec mcfly_prompt_command
```